[Regression] 5.7 – "cannot reference invalid declaration" error when accessing property wrapper inside closure

**Describe the bug**
Accessing a property wrapper using the underscore notation in a closure does not compile with Swift 5.7 (Xcode 14 RC).

This code reproduces the issue:

```swift
@propertyWrapper
struct Wrapper {
    var wrappedValue: Int
    
    init(wrappedValue: Int) {
        self.wrappedValue = wrappedValue
    }
}

class Foo {
    let bar: () -> Void = {
        @Wrapper var wrapped = 1
        
        let wrapper = _wrapped // compiles on 5.6, errors out on 5.7
    }
}

```

**Steps To Reproduce**
1. Add the code above to a Wrapper.swift file
2. Compile the file with the Swift 5.7 compiler: `swiftc Wrapper.swift`
3. The following compilation error occurs: 
```
❯ swiftc Wrapper.swift
Wrapper.swift:14:23: error: cannot reference invalid declaration '_wrapped'
        let wrapper = _wrapped // compiles on 5.6, errors out on 5.7
                      ^
Wrapper.swift:12:22: note: '_wrapped' declared here
        @Wrapper var wrapped = 1
```


**Expected behavior**
Accessing the property wrapper using the underscore notation code should compile, as it does on Swift 5.6.


**Environment**
 - Xcode 14 RC toolchain


**Additional context**
The same code produces a SILGen crash with the 5.7 nightly toolchain: https://godbolt.org/z/o1qGnov7o

```
unsupported collection upcast kind
UNREACHABLE executed at /home/build-user/swift/lib/SILGen/SILGenPoly.cpp:523!
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/swift-5.7/usr/bin/swift-frontend -frontend -S -primary-file <source> -target x86_64-unknown-linux-gnu -disable-objc-interop -g -new-driver-path /opt/compiler-explorer/swift-5.7/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /opt/compiler-explorer/swift-5.7/usr/lib/swift -enable-anonymous-context-mangled-names -Xllvm --x86-asm-syntax=intel -module-name output -o /app/output.s
1.	Swift version 5.7.1-dev (LLVM 7e31b78ab3255fa, Swift 1e80674d67f5505)
2.	Compiling with the current language version
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "<source>")
4.	While silgen emitStoredPropertyInitialization SIL function "@$s6output3FooC3baryycvpfi".
 for expression at [<source>:11:27 - line:15:5] RangeText="{
        @Wrapper var wrapped = 1
        
        let wrapper = _wrapped // works on 5.6, crashes on 5.7
    "
5.	While silgen closureexpr SIL function "@$s6output3FooC3baryycvpfiyycfU_".
 for expression at [<source>:11:27 - line:15:5] RangeText="{
        @Wrapper var wrapped = 1
        
        let wrapper = _wrapped // works on 5.6, crashes on 5.7
    "

...

Compiler returned: 139
```

--- a/lib/Sema/CSClosure.cpp
+++ b/lib/Sema/CSClosure.cpp
@@ -530,18 +530,13 @@ private:
       init = TypeChecker::buildDefaultInitializer(patternType);
     }
 
-    if (init) {
-      return SolutionApplicationTarget::forInitialization(
-          init, patternBinding->getDeclContext(), patternType, patternBinding,
-          index,
-          /*bindPatternVarsOneWay=*/false);
-    }
-
-    // If there was no initializer, there could be one from a property
-    // wrapper which has to be pre-checked before use. This is not a
-    // problem in top-level code because pattern bindings go through
-    // `typeCheckExpression` which does pre-check automatically and
-    // result builders do not allow declaring local wrapped variables.
+    // A property wrapper initializer (either user-defined
+    // or a synthesized one) has to be pre-checked before use.
+    //
+    // This is not a problem in top-level code because pattern
+    // bindings go through `typeCheckExpression` which does
+    // pre-check automatically and result builders do not allow
+    // declaring local wrapped variables (yet).
     if (hasPropertyWrapper(pattern)) {
       auto target = SolutionApplicationTarget::forInitialization(
           init, patternBinding->getDeclContext(), patternType, patternBinding,
@@ -554,6 +549,13 @@ private:
         return None;
 
       return target;
+    }
+
+    if (init) {
+      return SolutionApplicationTarget::forInitialization(
+          init, patternBinding->getDeclContext(), patternType, patternBinding,
+          index,
+          /*bindPatternVarsOneWay=*/false);
     }
 
     return SolutionApplicationTarget::forUninitializedVar(patternBinding, index,

--- a/lib/Sema/CSClosure.cpp
+++ b/lib/Sema/CSClosure.cpp
@@ -113,7 +113,7 @@ public:
           CS.setType(var, computeProjectedValueType(wrappedVar, wrapperType));
         } else {
           // _<name> is the wrapper var
-          CS.setType(var, computeWrappedValueType(wrappedVar, wrapperType));
+          CS.setType(var, wrapperType);
         }
 
         return {true, expr};

--- a/test/expr/closure/multi_statement.swift
+++ b/test/expr/closure/multi_statement.swift
@@ -566,3 +566,22 @@ struct Test {
     }
   }
 }
+
+https://github.com/apple/swift/issues/61017
+do {
+  @propertyWrapper
+  struct Wrapper {
+    var wrappedValue: Int
+
+    init(wrappedValue: Int) {
+      self.wrappedValue = wrappedValue
+    }
+  }
+
+  class Test {
+    let bar: () -> Void = {
+      @Wrapper var wrapped = 1
+      let wrapper: Wrapper = _wrapped // Ok
+    }
+  }
+}

--- a/test/expr/closure/multi_statement.swift
+++ b/test/expr/closure/multi_statement.swift
@@ -585,3 +585,44 @@ do {
     }
   }
 }
+
+https://github.com/apple/swift/issues/61024
+do {
+  enum Baz: String {
+  case someCase
+  }
+
+  @propertyWrapper
+  struct Wrapper {
+    var wrappedValue: Int
+    let argument: String
+
+    init(wrappedValue: Int, argument: String) { // expected-note 2 {{'init(wrappedValue:argument:)' declared here}}
+      self.wrappedValue = wrappedValue
+      self.argument = argument
+    }
+  }
+
+  class Foo {
+    let ok: () -> Void = {
+      @Wrapper(argument: Baz.someCase.rawValue) var wrapped1 = 1 // Ok
+      @Wrapper(wrappedValue: 42, argument: Baz.someCase.rawValue) var wrapped2 // Ok
+      @Wrapper(wrappedValue: 42, argument: Baz.someCase.rawValue) var wrapped3: Int // Ok
+    }
+
+    let bad0: () -> Void = {
+      @Wrapper var wrapped: Int
+      // expected-error@-1 {{missing arguments for parameters 'wrappedValue', 'argument' in call}}
+    }
+
+    let bad1: () -> Void = {
+      @Wrapper var wrapped = 0
+      // expected-error@-1 {{missing argument for parameter 'argument' in property wrapper initializer; add 'wrappedValue' and 'argument' arguments in '@Wrapper(...)'}}
+    }
+
+    let bad2: () -> Void = {
+      @Wrapper(wrappedValue: 42, argument: Baz.someCase.rawValue) var wrapped = 0
+      // expected-error@-1 {{extra argument 'wrappedValue' in call}}
+    }
+  }
+}

--- a/validation-test/Sema/type_checker_crashers_fixed/issue59294.swift
+++ b/validation-test/Sema/type_checker_crashers_fixed/issue59294.swift
@@ -74,7 +74,7 @@ class Test {
       if $value.returnValue() > 0 {
         test {
           return $value.returnValue() == $0 &&
-                 _value == $0
+                 _value.wrappedValue == $0
         }
       }
       return false
@@ -84,7 +84,7 @@ class Test {
       @WrapperValue var value = $0
 
       test {
-        if _value != $0 { // Ok
+        if _value.wrappedValue != $0 { // Ok
           $value.printValue()
         }
       }
@@ -99,7 +99,8 @@ class Test {
       if true {
         @OuterWrapper @WrapperValue var value = $0
         if true {
-          let _: Bool = _value == $0
+          let _: OuterWrapper<WrapperValue<Bool>> = _value
+          let _ = _value.wrappedValue.wrappedValue == $0
           let _: OuterWrapper<WrapperValue<Bool>> = $value
           let _: Bool = value
         }

